### PR TITLE
Refactored code to remove unnecessary checks

### PR DESCRIPTION
**Pull request by pi1024e**
When creating an immutable funnel, the code checks the value of "hasLengthByte" twice if it ends up being true. Checking for it once is all that is needed, as another branch can check for the other boolean statement in the if statement previously on line 1483. Differentiating the branches allows less redundant checks and does not introduce any side effects.

### CODE REVIEW

**Comment by pi1024e** [comment]
@swift-ci please test linux

**Comment by pi1024e** [comment]
@swift-ci please test

File: CoreFoundation/String.subproj/CFString.c [hunk 1]
```diff
@@ -1480,12 +1480,14 @@ CF_PRIVATE CFStringRef __CFStringCreateImmutableFunnel3(
                 useInlineData = true;
                 size = numBytes;
 
-                if (hasLengthByte || (encoding != kCFStringEncodingUnicode && __CFCanUseLengthByte(numBytes))) {
+                if (hasLengthByte) {
+                    useLengthByte = true;
+                    size += 1;
```
Comment (gregomni) [inline CoreFoundation/String.subproj/CFString.c]: It looks like you've reversed which condition gets the +1 size here? Shouldn't it be in !hasLengthByte?

**Comment by pi1024e** [comment]
Fixed the mistake. Everything should be good!

**Comment by pi1024e** [comment]
@swift-ci please test

**Comment by pi1024e** [comment]
@swift-ci please test linux

**Comment by pi1024e** [comment]
cc @gregomni

**Comment by xwu** [comment]
@pi1024e Could you please squash this into one commit?

**Comment by xwu** [comment]
@swift-ci Please test Linux

**Comment by pi1024e** [comment]
@swift-ci Please test linux

**Comment by millenomi** [comment]
@swift-ci Please test linux

File: CoreFoundation/String.subproj/CFString.c [hunk 2]
```diff
@@ -5139,9 +5139,9 @@ void CFStringLowercase(CFMutableStringRef string, CFLocaleRef locale) {
 
     if (!langCode && isEightBit) {
         uint8_t *contents = (uint8_t *)__CFStrContents(string) + __CFStrSkipAnyLengthByte(string);
-        for (;currentIndex < length;currentIndex++) {
+        while (currentIndex < length) {
             if (contents[currentIndex] >= 'A' && contents[currentIndex] <= 'Z') {
-                contents[currentIndex] += 'a' - 'A';
+                contents[currentIndex++] += 'a' - 'A';
```
Comment (drodriguez) [inline CoreFoundation/String.subproj/CFString.c]: I don't think this change should be merged. It is not completely equivalent in the case neither the `if` branch, nor the `else if` branch are taken, and will create an infinite loop.

File: CoreFoundation/String.subproj/CFString.c [hunk 3]
```diff
@@ -4669,7 +4667,7 @@ CFDataRef CFStringCreateExternalRepresentation(CFAllocatorRef alloc, CFStringRef
 	// if result == length, we always succeed
 	//   otherwise, if result == 0, we fail
 	//   otherwise, if there was a lossByte but still result != length, we fail
-        if ((result != length) && (!result || !lossByte)) return NULL;
+        if ((result != length) && !(result && lossByte)) return NULL;
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: This logic is harder to follow. The comments above explain the thought process and the existing code is reasonably readable, while `result && lossByte` is not.

**Comment by pi1024e** [comment]
Made those changes!

File: CoreFoundation/String.subproj/CFString.c [hunk 4]
```diff
@@ -4667,7 +4667,7 @@ CFDataRef CFStringCreateExternalRepresentation(CFAllocatorRef alloc, CFStringRef
 	// if result == length, we always succeed
 	//   otherwise, if result == 0, we fail
 	//   otherwise, if there was a lossByte but still result != length, we fail
-        if ((result != length) && (!result || !lossByte)) return NULL;
+        if ((result != length) && !result || !lossByte)) return NULL;
```
Comment (drodriguez) [inline CoreFoundation/String.subproj/CFString.c]: This is missing a parenthensis before `!result`

File: CoreFoundation/String.subproj/CFString.c [hunk 5]
```diff
@@ -5139,12 +5139,13 @@ void CFStringLowercase(CFMutableStringRef string, CFLocaleRef locale) {
...
     if (!langCode && isEightBit) {
         uint8_t *contents = (uint8_t *)__CFStrContents(string) + __CFStrSkipAnyLengthByte(string);
-        for (;currentIndex < length;currentIndex++) {
+        while (currentIndex < length) {
             if (contents[currentIndex] >= 'A' && contents[currentIndex] <= 'Z') {
                 contents[currentIndex] += 'a' - 'A';
             } else if (contents[currentIndex] > 127) {
                 break;
             }
+	currentIndex++;
```
Comment (drodriguez) [inline CoreFoundation/String.subproj/CFString.c]: Looks to not be indented to the right level.

While I understand that a `for` loop without the initialization condition looks weird, I still think it shows the intent better than a `while` loop and increasing the index inside the loop.

File: CoreFoundation/String.subproj/CFString.c [hunk 6]
```diff
@@ -491,6 +489,7 @@ CF_INLINE Boolean __CFStrEncodingCanBeStoredInEightBit(CFStringEncoding encoding
         case kCFStringEncodingInvalidId:
         case kCFStringEncodingUnicode:
         case kCFStringEncodingNonLossyASCII:
+	default:
```
Comment (drodriguez) [inline CoreFoundation/String.subproj/CFString.c]: Incorrect indentation?

Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Moving `default` up is a very strange coding style. I don't think we should make this change.

File: CoreFoundation/String.subproj/CFString.c [hunk 7]
```diff
@@ -5466,8 +5462,8 @@ void CFStringNormalize(CFMutableStringRef string, CFStringNormalizationForm theF
 
         contents = (uint8_t *)__CFStrContents(string) + __CFStrSkipAnyLengthByte(string);
 
-        for (;currentIndex < length;currentIndex++) {
-            if (contents[currentIndex] > 127) {
+        while (currentIndex < length;currentIndex++) {
+            if (contents[currentIndex++] > 127) {
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: This is now double incremented? The change is not correct, or it is misleading to read.

**Comment by pi1024e** [comment]
Made those changes. Thank you!

File: CoreFoundation/String.subproj/CFString.c (hunk 3 above)
Comment (xwu) [reply CoreFoundation/String.subproj/CFString.c]: This has not been resolved; please revert to the original.

File: CoreFoundation/String.subproj/CFString.c (hunk 5 above)
Comment (xwu) [reply CoreFoundation/String.subproj/CFString.c]: This also has not been resolved. Please revert this loop to the original version.

File: CoreFoundation/String.subproj/CFString.c [hunk 8]
```diff
@@ -5466,12 +5463,13 @@ void CFStringNormalize(CFMutableStringRef string, CFStringNormalizationForm theF
 
         contents = (uint8_t *)__CFStrContents(string) + __CFStrSkipAnyLengthByte(string);
 
-        for (;currentIndex < length;currentIndex++) {
+        while (currentIndex < length) {
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Likewise here.

File: CoreFoundation/String.subproj/CFString.c [hunk 9]
```diff
@@ -5157,9 +5158,7 @@ void CFStringLowercase(CFMutableStringRef string, CFLocaleRef locale) {
 
         if (isEightBit) __CFStringChangeSize(string, CFRangeMake(0, 0), 0, true);
 
-        contents = (UniChar *)__CFStrContents(string);
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Please also revert this to the original version.

File: CoreFoundation/String.subproj/CFString.c [hunk 10]
```diff
@@ -5248,9 +5247,7 @@ void CFStringUppercase(CFMutableStringRef string, CFLocaleRef locale) {
...
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Please also revert this change.

File: CoreFoundation/String.subproj/CFString.c [hunk 11]
```diff
@@ -5713,8 +5711,8 @@ void CFStringNormalize(CFMutableStringRef string, CFStringNormalizationForm theF
                         if (CFUniCharIsSurrogateHighCharacter(nextChar) && ((contents + 1) < limit) && CFUniCharIsSurrogateLowCharacter(*(contents + 1))) {
                             nextChar = CFUniCharGetLongCharacterForSurrogatePair(nextChar, *(contents + 1));
                             if (!CFUniCharIsMemberOfBitmap(nextChar, (const uint8_t *)CFUniCharGetBitmapPtrForPlane(kCFUniCharNonBaseCharacterSet, (nextChar >> 16))) || (0 != CFUniCharGetCombiningPropertyForCharacter(nextChar, (const uint8_t *)CFUniCharGetUnicodePropertyDataForPlane(kCFUniCharCombiningProperty, (nextChar >> 16))))) break;
-                        } else {
-                            if (!CFUniCharIsMemberOfBitmap(nextChar, nonBaseBMP) || (0 != CFUniCharGetCombiningPropertyForCharacter(nextChar, combiningBMP))) break;
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: The logic was more clear in the original version; the `if` statement here parallels that above. Please revert to the original.

**Comment by pi1024e** [comment]
Fixed!

File: CoreFoundation/String.subproj/CFString.c [hunk 12]
```diff
@@ -5139,7 +5139,7 @@ void CFStringLowercase(CFMutableStringRef string, CFLocaleRef locale) {
 
     if (!langCode && isEightBit) {
         uint8_t *contents = (uint8_t *)__CFStrContents(string) + __CFStrSkipAnyLengthByte(string);
-        for (;currentIndex < length;currentIndex++) {
+        for (;currentIndex < length; currentIndex++) {
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Please avoid making unrelated whitespace changes and restore the original code.

File: CoreFoundation/String.subproj/CFString.c [hunk 13]
```diff
@@ -5156,9 +5156,7 @@ void CFStringLowercase(CFMutableStringRef string, CFLocaleRef locale) {
         UInt32 flags = 0;
 
         if (isEightBit) __CFStringChangeSize(string, CFRangeMake(0, 0), 0, true);
-
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Likewise here.

File: CoreFoundation/String.subproj/CFString.c [hunk 14]
```diff
@@ -5156,9 +5156,7 @@ void CFStringLowercase(CFMutableStringRef string, CFLocaleRef locale) {
...
         contents = (UniChar *)__CFStrContents(string);
-
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: And here.

File: CoreFoundation/String.subproj/CFString.c (hunk 10 above)
Comment (xwu) [reply CoreFoundation/String.subproj/CFString.c]: This is not correctly resolved. It is missing a semicolon and won't compile. There is also an unnecessary whitespace change.

File: CoreFoundation/String.subproj/CFString.c [hunk 15]
```diff
@@ -5466,7 +5463,7 @@ void CFStringNormalize(CFMutableStringRef string, CFStringNormalizationForm theF
 
         contents = (uint8_t *)__CFStrContents(string) + __CFStrSkipAnyLengthByte(string);
 
-        for (;currentIndex < length;currentIndex++) {
+        for (;currentIndex < length; currentIndex++) {
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Please also revert this unrelated whitespace change.

**Comment by drodriguez** [comment]
@pi1024e: I don't know what your expertise level is, but I'm going to explain this as best as I can.

The only commit that we are currently interested is the first one, and also the fix from the first feedback. It seems that you are using the `master` branch for working, I would recommend the following procedure (the lines that start in # are comments by me, the lines that start with $ will have to be typed in the command line).

```
# Let's save all your work into a local branch
$ git branch cleanup-save
# Let's reset your master branch to the first commit
$ git reset --hard 8611fd80328d30e7fce86ba529152f4240a5aa41
# At this point edit the file CFString.c to move the line size += 1 to the bottom branch. Save your edit and don't modify anything else.
# I will be good to verify that only those changes are being sent.
$ git diff
# After checking that the changes are only the ones we want (the move of the line size += 1), we can continue.
$ git add CoreFoundation/String.subproj/CFString.c
# Let's rewrite the commit to have the fix.
$ git commit --amend --no-edit
# And finally update the remote to that version. I imagine your remote might be called origin, but replace <your remote name> in the line below by the name of your remote.
$ git push -f <your remote name>
```

That should get us only the initial changes, clean of every other commit. All your other changes are in a local branch called `cleanup-save`, and we can discuss about then in different PRs.

Thanks!

**Comment by pi1024e** [comment]
Thank you so much! Now I know what to do and how to properly contribute. Thank you!

File: CoreFoundation/String.subproj/CFString.c (hunk 13 above)
Comment (xwu) [reply CoreFoundation/String.subproj/CFString.c]: This needs to be reverted.

File: CoreFoundation/String.subproj/CFString.c [hunk 16]
```diff
@@ -5249,7 +5249,7 @@ void CFStringUppercase(CFMutableStringRef string, CFLocaleRef locale) {
         if (isEightBit) __CFStringChangeSize(string, CFRangeMake(0, 0), 0, true);
 
         contents = (UniChar *)__CFStrContents(string);
-
+
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: This whitespace change is new and needs to be reverted.

File: CoreFoundation/String.subproj/CFString.c (hunk 15 above)
Comment (xwu) [reply CoreFoundation/String.subproj/CFString.c]: This change still needs to be reverted.

**Comment by pi1024e** [comment]
Clang-format automatically made the whitespace changes. I thought I had to run that before committing to ensure no formatting changes would be done. Guess I was wrong.

**Comment by xwu** [comment]
It can be helpful to clang-format your own contributions, but you should not format existing code.

There are now pervasive changes in whitespace which cannot be taken as part of this PR. Again, please try to follow @drodriguez’s instructions _as exactly as possible_ and make _no other changes_.

**Comment by pi1024e** [comment]
Everything is fixed!

**Comment by xwu** [comment]
This looks good. As the final step, please squash your commits into one commit.

Here is a tutorial that explains how:
https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

**Comment by pi1024e** [comment]
Squashed those commits!

**Comment by spevans** [comment]
@swift-ci test

**Comment by pi1024e** [comment]
Are these changes ok?

**Comment by xwu** [comment]
This looks good to me, but others still have to approve.

**Comment by xwu** [comment]
@swift-ci Please test

**Comment by pi1024e** [comment]
@swift-ci please test

**Comment by pi1024e** [comment]
Fixed!

File: CoreFoundation/String.subproj/CFString.c [hunk 17]
```diff
@@ -1476,16 +1476,17 @@ CF_PRIVATE CFStringRef __CFStringCreateImmutableFunnel3(
                 useNullByte = hasNullByte;
 
             } else {	// Inline data; reserve space for it
-
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: Again, please don't make unrelated whitespace changes. After reverting this change, please squash the commit.

File: CoreFoundation/String.subproj/CFString.c [hunk 18]
```diff
@@ -1476,16 +1476,18 @@ CF_PRIVATE CFStringRef __CFStringCreateImmutableFunnel3(
...
+
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c]: This is bleeding whitespace that wasn't there before. Please be careful to revert the commit so that you don't unintentionally introduce such changes.

**Comment by xwu** [comment]
Good, now if you squash the commit, this PR looks to be in better shape.

**Comment by pi1024e** [comment]
Fixed!

**Comment by pi1024e** [comment]
@swift-ci please test

File: CMakeLists.txt [hunk 19]
```diff
@@ -7,7 +7,7 @@ list(APPEND CMAKE_MODULE_PATH
 # NOTE(compnerd) enable CMP0091 - select MSVC runtime based on
 # CMAKE_MSVC_RUNTIME_LIBRARY.  Requires CMake 3.15 or newer
 if(POLICY CMP0091)
-  cmake_policy(CMP0091 NEW)
+  cmake_policy(SET CMP0091 NEW)
```
Comment (xwu) [inline CMakeLists.txt]: What happened here? This is an unrelated change.

File: CoreFoundation/String.subproj/CFString.c (hunk 18 above)
Comment (xwu) [reply CoreFoundation/String.subproj/CFString.c]: It appears this bleeding whitespace is back. Please revert.

**Comment by pi1024e** [comment]
Done!

File: CMakeLists.txt (hunk 19 above)
Comment (pi1024e) [reply CMakeLists.txt]: Squashing a rebase caused this to happen. I have no idea how.

**Comment by pi1024e** [comment]
Okay. All changes done! Let us get this over with!

**Comment by pi1024e** [comment]
@swift-ci Please test

**Comment by xwu** [comment]
You need permissions to trigger testing. It's helpful if you've made sure that your changes compile and pass all tests locally.

@swift-ci Please test

**Comment by xwu** [comment]
@swift-ci Please test

**Comment by xwu** [comment]
@swift-ci Please test Linux
